Guidance: Implement the new mobile master bar

alshakero opened the pull request:
#### Changes proposed in this Pull Request

* In preparation for adding the help center entry point to the masterbar, this PR tucks reader, publish, and profile links in a menu.

This shouldn't change anything in the Desktop experience.

#### Testing instructions

1. In Chrome desktop, go to http://calypso.localhost:3000/
2. Once done loading, open DevTools and simulate mobile.
3. You should see a popover promoting the new navbar.
4. Clicking OK should hide the popover.
5. A menu button should appear and reader, publish, and profile link should be inside it.

I added a don't merge label until we figure persisting the user response and wait for Poli's feedback.

#### Screenshot
<kbd><img width="300" src="https://user-images.githubusercontent.com/17054134/162798985-7974d3ab-f0b6-4022-a9ee-3b76ed456dc8.png"></kbd>

alshakero committed:
Implement the new mobile master bar

escapemanuele commented on client/layout/masterbar/logged-in.jsx:
```diff
@@ -230,111 +248,186 @@ class MasterbarLoggedIn extends Component {
+				>
+					{ translate( 'Upgrade' ) }
+				</AsyncLoad>
+			),
```
Maybe the planUpsell needs the `! domainOnlySite && ! isMigrationInProgress` check as publish?

escapemanuele commented on client/layout/masterbar/logged-in.jsx:
```diff
@@ -230,111 +248,186 @@ class MasterbarLoggedIn extends Component {
+						{ translate( 'Search Actions' ) }
+					</Item>
+				),
+			planUpsell: () => (
+				<AsyncLoad
+					require="./plan-upsell"
+					className="masterbar__item-upsell button is-primary"
+					tooltip={ translate( 'Upgrade your plan' ) }
+				>
+					{ translate( 'Upgrade' ) }
+				</AsyncLoad>
+			),
```
Maybe the planUpsell needs the `! domainOnlySite && ! isMigrationInProgress` check the same as publish?

alshakero replied:
Phew, great catch.

escapemanuele reviewed (changes requested):
Wow! Your refactor of the masterbar is really astonishing, I really like it.
I've only a little doubt that I explained in the comment below!

alshakero committed:
Address feedback and implement the popover

alshakero commented:
Thanks for the review! I addressed your feedback and implemented the popover. 

Design for reference 
![image](https://user-images.githubusercontent.com/17054134/162420641-66f0de4d-39e5-4930-beec-99285ed83fbb.png)

alshakero committed:
Implment storing the popover response

alshakero committed:
Use user display instead of "My profile" on mobile

poligilad-auto commented:
Hey @alshakero,
Looking really cool. 

I have a few initial comments:
1. **The popover** is sticking to the right side of the screen. I assume this is because it's pointing to the "hamburger" icon. Could we have it point elsewhere to have some nice spacing from the edge of the screen? 16px would be ideal.
<img width="430" alt="image" src="https://user-images.githubusercontent.com/83961704/162456653-fb27d8f5-bde3-4774-b290-2aa525ab233d.png">


2. **Interaction**: we should close the "hamburger menu" whenever navigating elsewhere. Currently it stays open when I click on it > click on "notifications" / click on "write" etc.
3. **Width of the menu items**: The "notifications" and "cart" are 48px and the "hamburger menu" & "W" are 40px. They should all be a wider 48px. It's easier to click on mobile.
<img width="72" alt="image" src="https://user-images.githubusercontent.com/83961704/162457695-ce3602f9-1921-42bb-a53a-864bb0e3f368.png"> 
<img width="129" alt="image" src="https://user-images.githubusercontent.com/83961704/162458195-f433c96d-9267-454f-91bc-9c309a0b9ada.png">

4. **"Notification" alignment issue**: There seems to be an issue when the bell doesn't have the dot it's not aligned to the middle of the div. Can we fix that?
5. **I just updated the design of the items in the "hamburger menu"**. So let's make those changes. The items are taking font sizes and colors from the left nav bar menu on mobile. All font sizes are 16px and there are some changes needed for the "write" button. Checkout the Figma for specifics and let me know if you have questions.
<img width="351" alt="image" src="https://user-images.githubusercontent.com/83961704/162459608-4e5a6f64-f1df-4726-b9c0-99d893de1abe.png">
6. **The editor menu** should be much simpler.
<img width="351" alt="image" src="https://user-images.githubusercontent.com/83961704/162461251-4bcfb5a5-2200-4934-8b0e-c04ff5f64a0e.png">

alshakero committed:
Give the popover a margin on the right

alshakero committed:
Close the menu when the user navs or clicks away

tyxla commented on client/layout/masterbar/masterbar-menu.tsx:
```diff
@@ -0,0 +1,46 @@
+	open: boolean;
+	onClose: () => void;
+} > = ( { children, open, onClose } ) => {
+	const currentURL = page.current;
```
Do we really need to pull the entire `page` module just to fetch the current path? Could we use something built-in, like `window.location.pathname` for example?

tyxla commented on client/layout/masterbar/logged-in.jsx:
```diff
@@ -230,111 +259,218 @@ class MasterbarLoggedIn extends Component {
+						onClose={ () => this.handleToggleMenu() }
+						open={ this.state.isMenuOpen }
+					>
+						{ this.masterbarItemsMap().publish() }
```
What's the benefit of containing each item inside an object returned from `this. masterbarItemsMap()` when compared to just having a separate method for each item? Won't it be simpler to just have a separate method for each item?

tyxla reviewed (commented):
I haven't tested, nor thoroughly reviewed, but just wanted to drop a couple of drive-by questions.

poligilad-auto commented:
@alshakero another thing I noticed is that when I'm on "My home" and try to click the "W" it's not opening the left nav menu anymore.

alshakero committed:
Include the top bar in the dropshadow

alshakero committed:
Use redux dispatchers properly;

alshakero committed:
Address feedback

alshakero committed:
Minor CSS adjustments

alshakero commented:
Hi @poligilad-auto and @tyxla! Thanks a lot for the feedback. I addressed everything.

To see the popover again, please untick "dismissed-masterbar-collapsable-menu-popover" and click save after following these steps:

<img width="967" alt="image" src="https://user-images.githubusercontent.com/17054134/162797982-917d088d-48b1-4196-8a12-d07e70a72b12.png">

alshakero committed:
Fix publish button margin on mobile

tyxla commented on client/layout/masterbar/logged-in.jsx:
```diff
@@ -201,140 +223,272 @@ class MasterbarLoggedIn extends Component {
+		);
+	};
+	renderPopupSearch = () => {
+		const isWordPressActionSearchFeatureEnabled = config.isEnabled( 'wordpress-action-search' );
```
Nit, but I've generally found it more readable to `return` early if we won't render anything, rather than using a ternary. For example:

```
if( ! config.isEnabled( 'wordpress-action-search' ) || ! isActionSearchVisible ) {
	return null;
}
```

followed by the `return <AsyncLoad ...` bit.

tyxla commented on client/assets/stylesheets/_nav-unification.scss:
```diff
@@ -144,6 +145,30 @@
+
+		.masterbar__toggle-drafts.button.is-borderless {
+			background: #32373c;
+			color: #eeeeee;
```
We could use an existing variable here as well:

```suggestion
			color: var( --color-sidebar-text );
```

Generally, I'd suggest trying to reuse variables as much as possible instead of hardcoding colors.

tyxla commented on client/assets/stylesheets/_nav-unification.scss:
```diff
@@ -144,6 +145,30 @@
 	@include breakpoint-deprecated( '<480px' ) {
+		// New post button
+		.masterbar__item-new {
+			background: #32373c;
```
Or perhaps we should use `--color-sidebar-submenu-background` instead of hardcoding the color?

```suggestion
			background: var( --color-sidebar-submenu-background );
```

tyxla commented on client/layout/masterbar/logged-in.jsx:
```diff
@@ -30,11 +34,19 @@ import { getCurrentLayoutFocus } from 'calypso/state/ui/layout-focus/selectors';
+import { MasterBarMobileMenu } from './masterbar-menu';
 import Notifications from './notifications';
 
+const MENU_POPOVER_PREFERENCE_KEY = 'dismissed-masterbar-collapsable-menu-popover';
```
I think it would be useful to update the preferences schema to support this new key.

tyxla commented on client/layout/masterbar/logged-in.jsx:
```diff
@@ -201,140 +223,272 @@ class MasterbarLoggedIn extends Component {
-			currentSelectedSiteId,
-		} = this.props;
+	handleToggleMenu = () => {
+		this.setState( { isMenuOpen: ! this.state.isMenuOpen } );
```
Relying on `this.state` when using `this.setState()` can lead to potential bugs, since React batches updates and the state value might be different from the one we expect. [Quoting from React docs](https://reactjs.org/docs/react-component.html#setstate):

> `setState()` does not always immediately update the component. It may batch or defer the update until later. This makes reading `this.state` right after calling `setState()` a potential pitfall. Instead, use `componentDidUpdate` or a `setState` callback (`setState(updater, callback)`), either of which are guaranteed to fire after the update has been applied. If you need to set the state based on the previous state, read about the `updater` argument below.

Addressing this should be straightforward:

```suggestion
		this.setState( ( state ) => ( { isMenuOpen: ! state.isMenuOpen } ) );
```

tyxla commented on client/layout/masterbar/logged-in.jsx:
```diff
@@ -201,140 +223,272 @@ class MasterbarLoggedIn extends Component {
+		const { isCheckoutPending, previousPath, siteSlug, isJetpackNotAtomic, title } = this.props;
+		return (
+			<AsyncLoad
+				require="calypso/layout/masterbar/checkout.tsx"
```
Do we really need the `.tsx`? This was like that before, so feel free to leave it like that.

tyxla commented on client/layout/masterbar/logged-in.jsx:
```diff
@@ -201,140 +223,272 @@ class MasterbarLoggedIn extends Component {
+
+	renderLanguageSwitcher = () => {
+		return (
+			( this.props.isSupportSession || config.isEnabled( 'quick-language-switcher' ) ) && (
```
Same as above for early `return`-ing to make the code less nested and more readable.

I actually see this a few more times across the PR, so that comment applies for those other instances as well.

tyxla commented on client/layout/masterbar/logged-in.jsx:
```diff
@@ -201,140 +223,272 @@ class MasterbarLoggedIn extends Component {
 
-		if ( isCheckout || isCheckoutPending ) {
-			return (
+	checkout = () => {
```
I'd suggest keeping the naming scheme and prefixing every method that renders anything with `render*`.

tyxla commented on client/layout/masterbar/logged-in.jsx:
```diff
@@ -201,140 +223,272 @@ class MasterbarLoggedIn extends Component {
 
-		if ( isCheckout || isCheckoutPending ) {
-			return (
+	checkout = () => {
```
On a side note, is it necessary to use arrow functions for all those rendering methods? Or can they be regular old-school function-like methods?

alshakero replied:
I always prefer the classic way. But occasionally prefer the arrow way because it automatically binds the function to `this`. However I don't think any of these are event listeners to benefit from that.

tyxla commented on client/layout/masterbar/style.scss:
```diff
@@ -564,6 +640,14 @@ $autobar-height: 20px;
 		transform: scale3d( 1.15, 1.15, 1.15 );
 	}
 }
+@keyframes mobile-menu-fade-in {
```
We do have an existing `appear` animation already that does essentially the same:

https://github.com/Automattic/wp-calypso/blob/9f3432741d318601ae081858ab2de6f55a5263eb/client/assets/stylesheets/shared/_animation.scss#L6

tyxla reviewed (commented):
This is nice work 👍 

Left a few more nits as I glanced through it again. Not a blocking review at all, feel free to take or throw any of the suggestions 😉

poligilad-auto commented:
Nice work @alshakero 👏

Just a few more comments:
1. In the editor the nav bar should be very different. Like in the image below. So basically swap the "W" for the "back" Gridicon and hide the rest.
<img width="719" alt="image" src="https://user-images.githubusercontent.com/83961704/162928400-f9579e91-96cc-4898-97b2-c689073dcb7a.png">
2. Some small tweaks to the "Write" button: 

- for "write" & the "draft number" can we make the font size also 14px & font-weight: 400?
- for the "draft number" & the "chevron": could we change the padding left and right to be 16px? That way the area is bigger and easier to click on mobile. We do have a lot more space on mobile for this.

Re: the initial pop-over didn't have a chance to see it again to check the positioning.

alshakero committed:
Address feedback

alshakero committed:
Address styling feedback

alshakero commented:
> In the editor the nav bar should be very different. Like in the image below. So basically swap the "W" for the "back" Gridicon and hide the rest.

Since this only has the help center entry point, we won't be able to ship it until ready. But I implemented everything and we'll just need to enable some code.

> 2. Some small tweaks to the "Write" button:

Addressed all accounts.

> Re: the initial pop-over didn't have a chance to see it again to check the positioning.

You should be able to see it now once more. But I wonder if you missed the steps to see as many times as you like here:

To see the popover again, please untick "dismissed-masterbar-collapsable-menu-popover" and click save after following these steps:

<img width="967" alt="image" src="https://user-images.githubusercontent.com/17054134/162797982-917d088d-48b1-4196-8a12-d07e70a72b12.png">

alshakero committed:
Merge branch 'trunk' into add/new-mobile-menu-bar

poligilad-auto commented:
Hey @escapemanuele, the nav bar within all the editors still need to changed (site editor, page editor, post editor). They are still using the same nav bar as the other calypso pages.

Please see no. 1 from my previous comment. Let me know if you have any questions.


<img width="469" alt="image" src="https://user-images.githubusercontent.com/83961704/163144725-062d349a-ff67-47c9-8499-eddac9ebd0ce.png">
<img width="448" alt="image" src="https://user-images.githubusercontent.com/83961704/163144811-78bf0f69-47aa-4a83-86d6-f42c9050de6e.png">

escapemanuele committed:
changed wordpress icon to chevron left

escapemanuele committed:
Design fixes

escapemanuele committed:
Fix desktop

escapemanuele reviewed (approved):
After checking with @poligilad-auto, the changes are good to go!

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7226015

Thank you @alshakero for including a screenshot in the description! This is really helpful for our translators.

enejb commented:
Thanks for working on this. 

I found that when you create a brand new user that just created a brand new site you end up with a double notice on the editor.

See
<img width="300" src="https://user-images.githubusercontent.com/115071/163891291-2237fdfc-d609-4dc2-89fb-99b2b5706abd.PNG" />

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7226015).
